### PR TITLE
[WIP] Run FSharp.Formatting on FCS only

### PR DESCRIPTION
**Pull request by cloudRoutine**
- remove dependency on VFPT.Core
- upgrade to FSharp 4.1
- colorize more token types
- upgrade to netcore?

@matthid after this maybe we could work on polishing up the infrastructure for this repo, setting up the dotliquid lib, and making this generally easier to work with?

I know I need to adjust the tests since there are a lot more token kinds and I'm not sure that the changes I made are working quite right. It'll probably be a while before I get around to finishing this, I've got a lot of other things I'm working on atm

### CODE REVIEW

**Comment by matthid** [comment]
@cloudRoutine not sure what you mean.

regarding these changes: I generally like the move away from vfpt, because it will get rid of some weird errors when the api of fcs broke. However: Our lobrary users still have the same problem (like vfpt.core has). Really the problem is that fcs breaks its api all the time and all core fsharp projects need to fix it. 

Isn't the correct dependency Fsharp.editing for colorization?

Maybe it's easier if this pr contains no colorization changes and just the updates?

File: packages/FSharp.Formatting/FSharp.Formatting.fsx [hunk 1]
```diff
@@ -2,24 +2,23 @@
 // Standard NuGet or Paket location
 #I __SOURCE_DIRECTORY__
 #I "lib/net40"
-#r "./lib/net40/System.Web.Razor.dll"
+#r "../Microsoft.AspNet.Razor/lib/net45/System.Web.Razor.dll"
 // Force load
 if (typeof<System.Web.Razor.ParserResults>.Assembly.GetName().Version.Major <= 2) then
   failwith "Wrong System.Web.Razor Version loaded!"
 
+#r "../System.ValueTuple/lib/portable-net40+sl4+win8+wp8/System.ValueTuple.dll"
```
Comment (matthid) [inline packages/FSharp.Formatting/FSharp.Formatting.fsx]: Btw this dependency was removed in latest FCS ...

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 2]
```diff
@@ -59,7 +64,7 @@ module private Helpers =
     let defines = defines |> Option.map (fun (s:string) -> 
       s.Split([| ' '; ';'; ',' |], StringSplitOptions.RemoveEmptyEntries) |> List.ofSeq)
     // Create source tokenizer
-    let sourceTok = SourceTokenizer(defaultArg defines [], file)
+    let sourceTok =   FSharpSourceTokenizer(defaultArg defines [], file)
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: space :)

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 3]
```diff
@@ -152,19 +163,23 @@ type CodeFormatAgent() =
...
+                    ToolTipReader.tryFormatTip tooltip with 
               | Some(_) as res -> res
-              | _ when island.Length > 1 ->
-                  // Try to find some information about the last part of the identifier 
-                  let tip = checkResults.GetIdentTooltip(line + 1, token.LeftColumn + 2, lines.[line], [ processDoubleBackticks body ])
-                  Async.RunSynchronously tip |> Option.bind ToolTipReader.tryFormatTip
+              //| _ when island.Length > 1 ->
+              //    // Try to find some information about the last part of the identifier 
+              //    let tip = checkResults.GetIdentTooltip(line + 1, token.LeftColumn + 2, lines.[line], [ processDoubleBackticks body ])
+              //    Async.RunSynchronously tip |> Option.bind ToolTipReader.tryFormatTip
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: Just remove when this is no longer needed.

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 4]
```diff
@@ -181,87 +196,88 @@ type CodeFormatAgent() =
...
 // --------------------------------------------------------------------------------------
 
   // Create an instance of an InteractiveChecker (which does background analysis
   // in a typical IntelliSense editor integration for F#)
-  let languageService = LanguageService()
-  do languageService.SetCriticalErrorHandler(fun exn str1 str2 opts ->
-      FSharp.Formatting.Common.Log.errorf "Language Service Error (%s, %s, %A): %O" str1 str2 opts exn)
+  let fsChecker = FSharpChecker.Create()
+  //do fsChecker.SetCriticalErrorHandler(fun exn str1 str2 opts ->
+  //    FSharp.Formatting.Common.Log.errorf "Language Service Error (%s, %s, %A): %O" str1 str2 opts exn)
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: I never have seen this error message. Do we need something similar now. Or can we just remove this?

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 5]
```diff
@@ -181,87 +196,88 @@ type CodeFormatAgent() =
...
-      let! checkResults = languageService.ParseAndCheckFileInProject(opts, file, source, AllowStaleResults.No)
-      let! symbolUses = languageService.GetAllUsesOfAllSymbolsInFile (opts, file, source, AllowStaleResults.No, false, new Profiler())
-      return checkResults, symbolUses
-  }
+  //let getTypeCheckInfo(filePath, source, opts) = asyncMaybe {
+      
+
+  //    let! symbolUses = checkResults.GetAllUsesOfAllSymbolsInFile () |> liftAsync
+  //    return checkResults, symbolUses
+  //}
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: dito

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 6]
```diff
@@ -272,37 +288,28 @@ type CodeFormatAgent() =
...
     let opts = 
       match options with 
-      | Some(str:string) when not(String.IsNullOrEmpty(str)) -> 
+      | Some(str:string) when not(System.String.IsNullOrEmpty(str)) -> 
           { opts with OtherOptions = Helpers.parseOptions str }
       | _ -> opts
 
     // Run the second phase - perform type checking
-    let checkResults, symbolUses = getTypeCheckInfo(file, source, opts) |> Async.RunSynchronously
+    //let! (checkResults, symbolUses) = getTypeCheckInfo(file, source, opts)
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: comment

File: src/FSharp.CodeFormat/CodeFormatAgent.fs [hunk 7]
```diff
@@ -358,8 +365,13 @@ type CodeFormatAgent() =
       // Receive parameters for the next parsing request
       let! request, (chnl:AsyncReplyChannel<_>) = agent.Receive()
       try
-        let! res, errs = processSourceCode request
-        chnl.Reply(Choice1Of2(res |> Array.ofList, errs))
+        //let! res, errs = processSourceCode request
```
Comment (matthid) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: comment

File: src/FSharp.CodeFormat/HtmlFormatting.fs [hunk 8]
```diff
@@ -115,39 +121,48 @@ let rec formatTokenSpans (ctx:FormattingContext) = List.iter (function
...
+        | TokenKind.MutableVar    -> CSS.MutableVar    
+        | TokenKind.Printf        -> CSS.Printf        
+        | TokenKind.Escaped       -> CSS.Escaped       
+        | TokenKind.Disposable    -> CSS.Disposable    
+        | TokenKind.TypeArgument  -> CSS.TypeArgument  
+        | TokenKind.Punctuation   -> CSS.Punctuation   
+        | TokenKind.Enumeration   -> CSS.Enumeration   
+        | TokenKind.Interface     -> CSS.Interface     
+        | TokenKind.Property      -> CSS.Property      
+        | TokenKind.UnionCase     -> CSS.UnionCase
```
Comment (matthid) [inline src/FSharp.CodeFormat/HtmlFormatting.fs]: I'm actually not a huge fan of this code style as it complicates adding new (longer) entries and makes the diff longer in such cases (and "blame" harder)...

But I'll probably accept without reverting this (Just for your info)

Comment (cloudRoutine) [inline src/FSharp.CodeFormat/HtmlFormatting.fs]: several of them are used in multiple places, they should not be hardcoded

**Comment by cloudRoutine** [comment]
you're gonna have to give up on the giant github diff for this PR as a whole being anything close to readable. Although you can always go through commit by commit, and then it's pretty easy to see the changes, that's also why i title organization and formatting commits as such so they can be skipped over easily, but everyone insists on doing it the hard way /smh

File: src/FSharp.CodeFormat/CodeFormatAgent.fs (hunk 5 above)
Comment (cloudRoutine) [inline src/FSharp.CodeFormat/CodeFormatAgent.fs]: i'm still getting all of the build configuration kinks worked out, i'll come back around and polish it up all spiffy once it actually works again 😉

File: src/FSharp.CodeFormat/HtmlFormatting.fs (hunk 8 above)
Comment (matthid) [inline src/FSharp.CodeFormat/HtmlFormatting.fs]: What I meant was the indentation of `->`. Extracting the strings is fine and good :)

**Comment by matthid** [comment]
Do we see comments in this PR when we comment in the commit?

File: src/FSharp.CodeFormat/HtmlFormatting.fs (hunk 8 above)
Comment (cloudRoutine) [inline src/FSharp.CodeFormat/HtmlFormatting.fs]: oh, hahaha. I have a hotkey that does the formatting, the main reason I format stuff like this is to do multi-line edits, keeping it all lined up and purrty doesn't really matter to me.

**Comment by matthid** [comment]
The other problem is that a review is there to verify that things are still good. If I skip half of the commits because they are called "paket update or whitespace changes" nobody verifies those commits and accidential problems might introduce 
 - for example some whitespaces matter in this repository
 - you might have wrongly commited other stuff in addition to whitespace changes
 - ...

File: build.fsx [hunk 9]
```diff
@@ -102,7 +102,7 @@ let solutionFile = "FSharp.Formatting.sln"
 
 let msbuild14 = ProgramFilesX86</>"MSBuild"</>"14.0"</>"Bin"</>"MSBuild.exe"
```
Comment (cloudRoutine) [inline build.fsx]: from inside the commit 😉

**Comment by cloudRoutine** [comment]
@matthid that's fair, but at that point you might as well just read through the source straight through once I'm done. 

There's a lot of technical debt that's accumulated here, and paying it off is gonna be messy 😁

**Comment by matthid** [comment]
> There's a lot of technical debt that's accumulated here, and paying it off is gonna be messy 😁

Sorry if I disturbed you, I thought it might help :). 
Ok I'll wait. Thanks for taking care of this

**Comment by cloudRoutine** [comment]
nah you didn't bother me, I just don't want you to waste your time perusing diffs that'll probably be outdated quickly.

**Comment by cloudRoutine** [comment]
@matthid apart from the 17 FSharp.Metadata.Format.Tests which are all failing due to YaafFSharpScripting this is the last failing test

```xml
Test Name:	Simple code snippet is formatted as HTML
Test FullName:	FSharp.CodeFormat.Tests.Simple code snippet is formatted as HTML
Test Source:	C:\Users\jared\Github\Forks\FSharp.Formatting\tests\FSharp.CodeFormat.Tests\Tests.fs : line 56
Test Outcome:	Failed
Test Duration:	0:00:00.104

Result StackTrace:	
at FsUnitTyped.TopLevelOperators.shouldContainText@62.Invoke(String message)
   at FsUnitTyped.TopLevelOperators.shouldContainText(String x, String y)
   at FSharp.CodeFormat.Tests.Simple code snippet is formatted as HTML() in C:\Users\jared\Github\Forks\FSharp.Formatting\tests\FSharp.CodeFormat.Tests\Tests.fs:line 57
Result Message:	
System.Exception : "<span class="k">let</span>" is not a substring of "<table class="pre"><tr><td class="lines"><pre class="fssnip"><span class="l">1: </span>
</pre></td>
<td class="snippet"><pre class="fssnip"><span class="m">let</span> <span onmouseout="hideTip(event, 'fstips1', 1)" onmouseover="showTip(event, 'fstips1', 1)" class="id">hello</span> <span class="o">=</span> <span class="n">10</span>
</pre></td>
</tr>
</table>"
```

**Comment by matthid** [comment]
There is still another problem with the C# test-project...

**Comment by cloudRoutine** [comment]
stupid xbuild. too many things don't work for it that are supposed to 😑

**Comment by cloudRoutine** [comment]
@matthid you can see the metadataformat errors that I was talking about earlier in the CI output now

File: build.fsx [hunk 10]
```diff
@@ -59,50 +59,53 @@ Target "AssemblyInfo" (fun _ ->
...
-    CleanDirs ["tests/FSharp.MetadataFormat.Tests/files/TestLib/bin"]
-    CleanDirs ["tests/FSharp.MetadataFormat.Tests/files/TestLib/obj"]
+    !! "bin"
+    ++ "temp"
+    ++ "docs/output"
+    ++ "tests/bin"
+    ++ "tests/FSharp.MetadataFormat.Tests/files/**/bin"
+    ++ "tests/FSharp.MetadataFormat.Tests/files/**/obj"
+    |> CleanDirs
+    CreateDir "tests/bin"
```
Comment (matthid) [inline build.fsx]: Something there is still not working...

File: .editorconfig [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+root = true
+
+[*]
+end_of_line = lf
+
+[*.{fs,fsi,fsx}]
+charset = utf-8
+indent_style = space
+indent_size = 4
+trim_trailing_whitespace = true
```
Comment (matthid) [inline .editorconfig]: Should this be gitignored? Which editor is this?

**Comment by cloudRoutine** [comment]
if you're trying to fix the travis build the issues are with stuff like 
```
Copying packages/FSharp.Core/lib/net45/FSharp.Core.dll to bin/FSharp.Core.dll
```
the paths aren't always full, but this isn't the real reason why stuff is failing
can you enable appveyor for this repo?

File: .editorconfig (hunk 11 above)
Comment (cloudRoutine) [inline .editorconfig]: it's supposed to be commited, it's for many many editors,

**Comment by matthid** [comment]
I can repro this on windows as well on a clean checkout...
But `CleanDirs` should create the directory if it doesn't exist. Something is wrong here

**Comment by cloudRoutine** [comment]
the paths need to be full, that's the problem

**Comment by matthid** [comment]
why?

**Comment by matthid** [comment]
It always worked like this?

**Comment by cloudRoutine** [comment]
¯\\\_(ツ)\_/¯
who knows?
Travis is the worst

**Comment by matthid** [comment]
I think the problem is that `!!` only matches existing stuff so its not the same as the old logic

**Comment by matthid** [comment]
so `!! "tests/bin"` is an empty sequence and `CleanDirs` never retrieves it as argument

File: .editorconfig (hunk 11 above)
Comment (matthid) [inline .editorconfig]: thanks didn't know that :)

**Comment by matthid** [comment]
`Associated XML file '/home/travis/build/tpetricek/FSharp.Formatting/tests/bin/csharpSupport.xml' was not found.`
Probably something with casing on linux... The defaults are `.XML` (no idea why)

**Comment by matthid** [comment]
This Build currently fails, because FAKE is using an older FSharp.Core and it therefore doesn't work on runtime... We should update FAKE to latest FSharp.Core (https://github.com/fsharp/FAKE/issues/1527).

This is only a quickfix, the correct fix is to update to the new netcore FAKE version which doesn't have this problem. Therefore I'll add a netcore output for FSF and change to the new FAKE version to get the version3  branch green.

**Pull request merged**